### ISSUE TITLE
Inner Join Error

### ISSUE THREAD

**Issue by nkim-radley** [Synergex/HarmonyCore#213]
Getting `System.NotImplementedException: The method or operation is not implemented.` error when attempting to $expand on a relationship.

RPS has relationship to and from files A and B both on key0. Expanding the relationship for either file:

`https://.../odata/v1/FileA?$expand=REL_FileB`

and

`https://.../odata/v1/FileB?$expand=REL_FileA`

... both produced the following error:

> fail: Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer[2]
> 
>       Connection ID "17798225733273780816", Request ID "80000251-0001-f700-b63f-84710c7967bb": An unhandled exception was thrown by the application.
> 
> System.NotImplementedException: The method or operation is not implemented.
> 
>    at Harmony.Core.EF.Query.Internal.HarmonyQueryExpression.AddInnerJoin(HarmonyQueryExpression innerQueryExpression, LambdaExpression outerKeySelector, LambdaExpression innerKeySelector, Type transparentIdentifierType)
> 
>    at Harmony.Core.EF.Query.Internal.HarmonyQueryableMethodTranslatingExpressionVisitor.TranslateJoin(ShapedQueryExpression outer, ShapedQueryExpression inner, LambdaExpression outerKeySelector, LambdaExpression innerKeySelector, LambdaExpression resultSelector)
> 
>    at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
> 
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
> 
>    at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
> 
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
> 
>    at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
> 
>    at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
> 
>    at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.AsyncEnumerableReader.ReadInternal[T](IAsyncEnumerable`1 value)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor.ExecuteAsyncEnumerable(ActionContext context, ObjectResult result, IAsyncEnumerable`1 asyncEnumerable)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
> 
> --- End of stack trace from previous location where exception was thrown ---
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
> 
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
> 
>    at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
> 
>    at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
> 
>    at Harmony.AspNetCore.Context.IMultiTenantMiddleware.$lc-553661668_InvokeMiddleware.MoveNext()
> 
> --- End of stack trace from previous location where exception was thrown ---
> 
>    at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
> 
>    at Microsoft.AspNetCore.Builder.Extensions.UsePathBaseMiddleware.Invoke(HttpContext context)
> 
>    at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()

Not sure if this is truly an error or if I am missing something in my regen.bat file.

**Comment by hippiehunter** [comment]
can you include any lines from your generated DBContext.dbl that look sort of like this?

```
parm.AddOneToOneRelation<FileA, FileB>("REL_FileA", "KEY_SOME_KEY", "KEY_OTHER_KEY")
```

**Comment by nkim-radley** [comment]
`parm.AddOneToOneToManyRelation<CashipHdr, CasystPar>("REL_CasystPar", "KEY_SHH_KEY0", "REL_CashipHdrs", "KEY_SP_KEY0")`

and

`parm.AddOneToOneToOneRelation<CasystPar, CadbsePar>("REL_CadbsePar", "KEY_SP_KEY0", "REL_CasystPar", "KEY_DB_KEY0")`

**Comment by hippiehunter** [comment]
I've fixed the issue with inner joins, but I don't think it should be happening at all. it looks like codegens defaults have picked a somewhat mismatched relation for those two types. One of them things its supposed to be one <-> one and the other thinks its supposed to be many <-> one. Do you currently have a harmony core customization json file in your templates folder?

**Comment by nkim-radley** [comment]
I do, but it was from that file that Steve generated for us. The customizations do include the following items:

<pre>
{
    "FromStructure": "CADBSE_PAR",  
    "FromKey": "DB_KEY0",  
    "ToStructure": "CASYST_PAR",  
    "ToKey": "SP_KEY0",  
    "RequiresMatch": true  
}</pre>

and

<pre>{
    "FromStructure": "CASYST_PAR",
    "FromKey": "SP_KEY0",
    "ToStructure": "CADBSE_PAR",
    "ToKey": "DB_KEY0",
    "RequiresMatch": true
}</pre>

**Comment by nkim-radley** [comment]
I updated our HarmonyCoreCustomization.json file and added the `RelationType` and `BackRelation` properties to both relationships (A<->B and B<->A). I'm no longer getting an HTTP 500 and getting a 200, but the expanded relation only shows "null."

DbContext.dbl after regen:
![image](https://user-images.githubusercontent.com/52933970/97033743-d1a89800-1531-11eb-9ba9-64efc05a24dd.png)
![image](https://user-images.githubusercontent.com/52933970/97033842-f69d0b00-1531-11eb-94a6-c5a10094cbb4.png)

HarmonyCoreCustomization.json
![image](https://user-images.githubusercontent.com/52933970/97033973-2a783080-1532-11eb-9ef2-3293ba6ff923.png)
![image](https://user-images.githubusercontent.com/52933970/97034001-37951f80-1532-11eb-8eab-6ed9735e6467.png)

**Comment by hippiehunter** [comment]
If you set DBG_SELECT=2 in your environment before running, it will dump debug info about your actual under the hood translated synergy select statement. Can you set that and run your query so we can see what it actually tried to do?

**Comment by nkim-radley** [comment]
I added DBG_SELECT and DBG_SELECT_FILE to my launch.bat and ran the same OData request using Postman, but I do not see the log file getting created. Is there a different location where I should be setting the DBG_SELECT environment variable?

**Comment by hippiehunter** [comment]
launch.bat is used to run your traditional synergy code. DBG_SELECT=2 needs to be set for your .net core hosting process. essentially, prior to running Services.Host.exe. You can set that in the debug tab for the Services.Host project or you could do a setlog/Environment.SetEnvironmentVariable inside your Services Custom file.

**Comment by nkim-radley** [comment]
Thank you for the explanation. I added the environment variables in the debug tab to my Services.Host project, I was getting HTTP 400 after building and publishing. I also tried running the solution from Visual Studio and am still getting HTTP 400 errors.

Here is what I set up:

![image](https://user-images.githubusercontent.com/52933970/97474761-d353d080-1922-11eb-909e-0c23e7db56ea.png)

**Comment by hippiehunter** [comment]
400 error sounds new, you started out getting 500's and then were getting 200's but the data was missing. are you sure you dont have something wrong with your query string? you should be getting 200 with missing data and something should be written to your log file

**Comment by nkim-radley** [comment]
In the case where I published the solution and hosted using IIS, I was getting HTTP 400 Bad Request when calling UserLogin (uses TraditionalBridge). Then when I tried to make Odata requests using Postman (with valid JWT), I was getting 401 Unauthorized.

When I hosted through Visual Studio, I only tried Odata requests and was getting connection refused errors.

**Comment by hippiehunter** [comment]
And the only change between success and failure is the addition of DBG_SELECT=2 and DBG_SELECT_FILE=c:\rweb\xfpl\dbg.log ? Do you have changes leftover in your launch.bat by any chance?

**Comment by nkim-radley** [comment]
After removing the environment variables from Services.Host, all my requests work again (i.e., no more http 400/401). I removed my leftover changes to launch.bat, but no changes in results.

**Comment by hippiehunter** [comment]
if you don't set DBG_SELECT_FILE, it will just log to the console. If your running from visual studio that should be pretty easy to see the result from. Can you just set DBG_SELECT=2 ? Overall this sounds pretty strange but I don't normally use DBG_SELECT_FILE for anything.

**Comment by nkim-radley** [comment]
Nevermind! Thank you for your patience, I was finally able to get the logging to work (I was using the wrong port number):


<pre>[Select debugging at :0]
 Implicit key selection via optimization: (See LRM)
   Key: Ordered nodup contig segs/Nodup contig segs/Ordered contig segs/Contig segs/Contig size
    >0: 0  1  0  1  31
From(C:\rweb\wglobal_data\WEBSYSFILES.DAT)
Optimized key: 0 (1:31) 
0000: C1 SEL_FLDREFDEF|SEL_FLDDEF|SEL_TYP_A
0001: 0000000A data length
0005: data: length=10
43 41 53 59 53 54 5F 50 41 52 
000F: 1F 02 SEL_OPC_FINDKEY(MEQ) krf = 0 (Head optimization)
      Optimized key range: 1:31
0011: 1D SEL_OPC_CDRETURN
0012: 1E SEL_OPC_STKRESET
0013: 17 SEL_OPC_READNEXT
0014: 13 SEL_OPC_IFERROR
0015: 00000011 opcode offset
0019: 81 SEL_FLDREFDEF|SEL_TYP_A
001A: 0000001F data length
001E: 00000000 data offset
0022: C1 SEL_FLDREFDEF|SEL_FLDDEF|SEL_TYP_A
0023: 0000000A data length
0027: data: length=10
43 41 53 59 53 54 5F 50 41 52 
0031: 03 SEL_OPC_EQ
0032: 12 SEL_OPC_IFTRUE (Tail optimization)
0033: 00000000 opcode offset
0037: 81 SEL_FLDREFDEF|SEL_TYP_A
0038: 0000001F data length
003C: 00000000 data offset
0040: C1 SEL_FLDREFDEF|SEL_FLDDEF|SEL_TYP_A
0041: 0000000A data length
0045: data: length=10
43 41 53 59 53 54 5F 50 41 52 
004F: 03 SEL_OPC_EQ
0050: 12 SEL_OPC_IFTRUE
0051: 00000012 opcode offset
0055: 20 SEL_OPC_SETEOF
0056: 1D SEL_OPC_CDRETURN
0057: 14 SEL_OPC_BRANCH
0058: 00000012 opcode offset
0033: 00000056 opcode offset update


[Select debugging at :0]
From(C:\rweb\cdat\CASYST_PAR.DAT) As a
Left Join From(c:\rweb\NQA\CADBSE_PAR.NQA) As b
On(b.A(1:2).EQ.a.A(1:2))
Non-optimized key: 0
      Driving table (a)
0000: 18 SEL_OPC_POSBOF krf = 0
0001: 1D SEL_OPC_CDRETURN
0002: 1E SEL_OPC_STKRESET
0003: 17 SEL_OPC_READNEXT
0004: 13 SEL_OPC_IFERROR
0005: 00000001 opcode offset
0009: 1D SEL_OPC_CDRETURN
000A: 14 SEL_OPC_BRANCH
000B: 00000002 opcode offset
      Inner table (b)
 Implicit key selection via optimization: (See LRM)
   Key: Ordered nodup contig segs/Nodup contig segs/Ordered contig segs/Contig segs/Contig size
    >0: 0  1  0  1  2
000F: A1 SEL_FLDREFDEF|SEL_RDXDEF|SEL_TYP_A
0010: 00000002 data length
0014: 00 record index
0015: 00000000 data offset
      Optimized key:0 with range: (1:2)
0019: 1F SEL_OPC_FINDKEY(MEQ) krf = 0 (Join optimization)
001B: 1E SEL_OPC_STKRESET
001C: 17 SEL_OPC_READNEXT
001D: 13 SEL_OPC_IFERROR
001E: 00000000 opcode offset
0022: A1 SEL_FLDREFDEF|SEL_RDXDEF|SEL_TYP_A
0023: 00000002 data length
0027: 01 record index
0028: 00000000 data offset
002C: A1 SEL_FLDREFDEF|SEL_RDXDEF|SEL_TYP_A
002D: 00000002 data length
0031: 00 record index
0032: 00000000 data offset
0036: 03 SEL_OPC_EQ
0037: 12 SEL_OPC_IFTRUE (Join full optimization)
0038: 00000000 opcode offset
003C: 20 SEL_OPC_SETEOF
003D: 1D SEL_OPC_CDRETURN
003E: 14 SEL_OPC_BRANCH
003F: 0000001B opcode offset
0038: 0000003D opcode offset update
001E: 0000003D opcode offset update</pre>

**Comment by hippiehunter** [comment]
it looks like its joining by matching the first 2 bytes of the record in CASYST_PAR.DAT and the first 2 bytes of the record in CADBSE_PAR.NQA. Is that the expected relationship?

**Comment by nkim-radley** [comment]
That is correct. And I confirmed both of those files have data and that there should be matches based on the first two bytes of both records (one to one to one relationship).

**Comment by hippiehunter** [comment]
Im going to build a small test for this, it might take a few hours.

**Comment by hippiehunter** [comment]
I've realized something while writing a test for this, are you sure you have the right relationship here? the first 2 characters are unique for both of these record types?

**Comment by hippiehunter** [comment]
I think i've reproduced the issue. Working now to try and solve/work around

**Comment by hippiehunter** [comment]
If you grab this updated template and copy it into your templates folder, then regen. I think it will solve your issue.  

https://github.com/Synergex/HarmonyCore/blob/master/Templates/ODataMetaData.tpl

you can see the specific changes i made here 

https://github.com/Synergex/HarmonyCore/commit/3169de89f9e19c3c87a1df09c65cce5291376a15#diff-6732fbfb2090ce1b889e7fc0d8fc645ec736a42e58abffe8413b9412b838fbf3

Basically I missed one of the relation cases in some recent work. We now have a test case for it so it should stay fixed.

**Comment by nkim-radley** [comment]
I updated the template, ran regen, but I'm now getting these errors for all models in my Services.Models project upon build:

<pre>
%DBL-E-NOOVR, No suitable Method found to override : public override method UpdateExistingJoin, void
</pre>

**Comment by hippiehunter** [comment]
For some reason I thought we had already updated you to the latest version but it looks like not. It might be easiest if you do the upgrade using the lastest command line tool, then replace the template.

**Comment by nkim-radley** [comment]
This issue seems to have been resolved with the update to 3.1.201 and forgot to close the issue. Thank you!
